prow/metrics/prowjobs/collector.go

```diff
@@ -157,3 +160,194 @@ func newHistogramVec() *prometheus.HistogramVec {
 		},
 	)
 }
+
+func newPlankProwJobHistogramVec() *prometheus.HistogramVec {
+	return prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "plank_pod_runtime_seconds",
+			Buckets: []float64{
+				time.Minute.Seconds() / 2,
+				(1 * time.Minute).Seconds(),
+				(2 * time.Minute).Seconds(),
+				(5 * time.Minute).Seconds(),
+				(10 * time.Minute).Seconds(),
+				(1 * time.Hour).Seconds() / 2,
+				(1 * time.Hour).Seconds(),
+				(2 * time.Hour).Seconds(),
+				(3 * time.Hour).Seconds(),
+				(4 * time.Hour).Seconds(),
+				(5 * time.Hour).Seconds(),
+				(6 * time.Hour).Seconds(),
+				(7 * time.Hour).Seconds(),
+				(8 * time.Hour).Seconds(),
+				(9 * time.Hour).Seconds(),
+				(10 * time.Hour).Seconds(),
+			},
+		},
+		[]string{
+			// namespace of the job
+			"job_namespace",
+			// name of the job
+			"job_name",
+			// type of the prowjob: presubmit, postsubmit, periodic, batch
+			"type",
+			// state of the prowjob pod: scheduled, initialized, run, postprocessed, succeeded, failed, unknown
+			"state",
+			// the org of the prowjob's repo
+			"org",
+			// the prowjob's repo
+			"repo",
+			// the base_ref of the prowjob's repo
+			"base_ref",
+		},
+	)
+}
+
+func updatePlank(histogramVec *prometheus.HistogramVec, jobInformer cache.SharedIndexInformer, oldPod *v1.Pod, newPod *v1.Pod) {
+	uid := newPod.ObjectMeta.Labels["prow.k8s.io/id"]
+	if uid == "" {
+		return
+	}
+	jobs, err := jobInformer.GetIndexer().Index("uid", uid)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to fetch prowJob from informer")
+		return
+	}
+	if len(jobs) == 0 {
+		return
+	}
+
+	if len(jobs) > 1 {
+		logrus.Error("Found more than one prowjob for pod")
+		return
+	}
+	job := jobs[0].(*prowapi.ProwJob)
+
+	if job.Status.PodName != newPod.Name {
+		// we don't care about this pod anymore
+		return
+	}
+
+	if newPod.DeletionTimestamp != nil {
+		// TODO, we will have to check for this on the ProwJob
+		return
+	}
+
+	oldState, oldTime := determineState(oldPod)
+	newState, newTime := determineState(newPod)
+
+	if oldState == newState {
+		return
+	}
+
+	labels := getPlankJobLabel(job, newState)
+	histogram, err := histogramVec.GetMetricWithLabelValues(labels.values()...)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to get a histogram for a prowjob")
+		return
+	}
+	histogram.Observe(newTime.Sub(oldTime.Time).Seconds())
+}
+
+func determineState(pod *v1.Pod) (state string, timestamp metav1.Time) {
```

Comment:
We have a couple missing bits here, the full workflow is

 - `clonerefs` to clone the repo
 - `initupload` to publish the result of the above
 - `place-tools` to inject `entrypoint` into the container
 - `test` that runs the test
 - `sidecar` to upload after it's done

Reply:
Will add them. I think I also have to take care of connection errors. When listwatcher loose the connection, I still have an older version in the cache when the connection is re-established, but then I can reconstruct the missed transitions, and can still be sure that I don't record them more than once.